packages/salesforcedx-vscode-core/src/commands/util/libraryCommandlet.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { Connection } from '@salesforce/core';
+import { ContinueResponse } from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+import { ToolingRetrieveResult } from '@salesforce/source-deploy-retrieve/lib/tooling/retrieve';
+import { ProgressLocation, window } from 'vscode';
+import { channelService } from '../../channels';
+import { nls } from '../../messages';
+import { notificationService } from '../../notifications';
+import { TelemetryData, telemetryService } from '../../telemetry';
+import { OrgAuthInfo } from '../../util';
+import { CommandletExecutor } from './sfdxCommandlet';
+
+export abstract class LibraryCommandletExecutor<T>
+  implements CommandletExecutor<T> {
+  protected showChannelOutput = true;
+  protected orgConnection: Connection | undefined;
+  protected executionName: string = '';
+  protected startTime: [number, number] | undefined;
+  protected telemetryName: string | undefined;
+
+  public execute(response: ContinueResponse<T>): void {}
+
+  public async build(
+    execName: string,
+    telemetryLogName: string
+  ): Promise<void> {
+    this.executionName = execName;
+    this.telemetryName = telemetryLogName;
+    // initialize connection
+    const usernameOrAlias = await OrgAuthInfo.getDefaultUsernameOrAlias(true);
+    if (!usernameOrAlias) {
+      throw new Error(nls.localize('error_no_default_username'));
+    }
+    this.orgConnection = await OrgAuthInfo.getConnection(usernameOrAlias);
+  }
+
+  public retrieveWrapper(
```

Comment:
This is where most of the visual aspects of the command execution are wired (publishing to output panel, showing progress message, etc). Initially I wanted to have it be generic enough to support calling the deploy, retrieve and other library methods but doing that wouldn't allow us to assume a response shape (line 58) so we could also have this format the results. After talking with Nick I opted to have a dedicated wrapper for the retrieve for now.